# Content-based recommendation system

In [1]:
import numpy as np 
import pandas as pd

import matplotlib.pyplot as plt

In [3]:
%matplotlib inline

## Log Data onto Weight&Bias

In [4]:
import wandb

In [5]:
DATA_PATH = './data/'

In [23]:
def log_dataset(use_artifact, 
        artifact_name, 
        artifact_path, dataset_name, 
        job_type='preprocess dataset', 
        project_name='Content-based RS'):

    run = wandb.init(project=project_name, job_type=job_type)
    run.use_artifact(use_artifact)
    artifact = wandb.Artifact(artifact_name, dataset_name)
    if os.path.isdir(artifact_path):
        artifact.add_dir(artifact_path)
    else:
        artifact.add_file(artifact_path)
    run.log_artifact(artifact)

## Data visualize

In [6]:
movie_df = pd.read_csv(os.path.join(DATA_PATH, 'ml-latest-small/movies.csv'))
movie_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


In [7]:
movie_df.head(10)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   
5        6                         Heat (1995)   
6        7                      Sabrina (1995)   
7        8                 Tom and Huck (1995)   
8        9                 Sudden Death (1995)   
9       10                    GoldenEye (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy  
5                        Action|Crime|Thriller  
6                               Comedy|Romance  
7                           Adventure|Children  
8                                       Action  
9                    Action|Adventure|Thriller

In [8]:
rate_df = pd.read_csv(os.path.join(DATA_PATH, 'ml-latest-small/ratings.csv'))
rate_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [9]:
rate_df.head(10)

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931
5       1       70     3.0  964982400
6       1      101     5.0  964980868
7       1      110     4.0  964982176
8       1      151     5.0  964984041
9       1      157     5.0  964984100

In [17]:
print('UNIQUE MOVIE ID:\t', len(movie_df['movieId'].unique()))
print('UNIQUE MOVIE TITLE:\t', len(movie_df['title'].unique()))

UNIQUE MOVIE ID:	 9742
UNIQUE MOVIE TITLE:	 9737


In `movies.csv`, there are **9742** unique movie Id, however, there are only **9737** unique title. Thus, some movies might have more than 1 Id.

In [10]:
movie_df[movie_df['title'].duplicated(keep=False)].sort_values(by='title')

movieId                                   title  \
4169     6003  Confessions of a Dangerous Mind (2002)   
9106   144606  Confessions of a Dangerous Mind (2002)   
650       838                             Emma (1996)   
5601    26958                             Emma (1996)   
5854    32600                             Eros (2004)   
9135   147002                             Eros (2004)   
2141     2851                         Saturn 3 (1980)   
9468   168358                         Saturn 3 (1980)   
5931    34048                War of the Worlds (2005)   
6932    64997                War of the Worlds (2005)   

                                   genres  
4169          Comedy|Crime|Drama|Thriller  
9106  Comedy|Crime|Drama|Romance|Thriller  
650                  Comedy|Drama|Romance  
5601                              Romance  
5854                                Drama  
9135                        Drama|Romance  
2141            Adventure|Sci-Fi|Thriller  
9468                      Sci-Fi|Thriller  
5931     Action|Adventure|Sci-Fi|Thriller  
6932                        Action|Sci-Fi

### Merge movies and ratings

In [12]:
df = pd.merge(rate_df, movie_df, on='movieId')

In [14]:
df.head(10)

userId  movieId  rating   timestamp             title  \
0       1        1     4.0   964982703  Toy Story (1995)   
1       5        1     4.0   847434962  Toy Story (1995)   
2       7        1     4.5  1106635946  Toy Story (1995)   
3      15        1     2.5  1510577970  Toy Story (1995)   
4      17        1     4.5  1305696483  Toy Story (1995)   
5      18        1     3.5  1455209816  Toy Story (1995)   
6      19        1     4.0   965705637  Toy Story (1995)   
7      21        1     3.5  1407618878  Toy Story (1995)   
8      27        1     3.0   962685262  Toy Story (1995)   
9      31        1     5.0   850466616  Toy Story (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1  Adventure|Animation|Children|Comedy|Fantasy  
2  Adventure|Animation|Children|Comedy|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy  
4  Adventure|Animation|Children|Comedy|Fantasy  
5  Adventure|Animation|Children|Comedy|Fantasy  
6  Adventure|Animation|Children|Comedy|Fantasy  
7  Adventure|Animation|Children|Comedy|Fantasy  
8  Adventure|Animation|Children|Comedy|Fantasy  
9  Adventure|Animation|Children|Comedy|Fantasy

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100836 entries, 0 to 100835
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
 4   title      100836 non-null  object 
 5   genres     100836 non-null  object 
dtypes: float64(1), int64(3), object(2)
memory usage: 7.4+ MB


In [19]:
# export to csv
df.to_csv(os.path.join(DATA_PATH, 'joined_movies.csv'))

In [24]:
# log to wandb
log_dataset(use_artifact='MovieLens-small:v0', 
            artifact_name='MovieLens-joined',
            artifact_path=os.path.join(DATA_PATH, 'joined_movies.csv'),
            dataset_name='PREPROCESS_DATASET')

wandb: wandb version 0.10.32 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
